### PR TITLE
Generify AttributeRenderer

### PR DESCRIPTION
**Pull request by Clashsoft**
Add a generic type variable on the AttributeRenderer interface and make methods using it generic, too.

This allows subclasses like NumberRenderer or StringRenderer to not have to use casts.

The change retains full binary compatibility because of erasure, and source compatibility for valid uses. Code like

```java
group.registerRenderer(Object.class, new StringRenderer());
```

now produces a compiler error. It would previously have failed at runtime due to the forced (String)o cast.

### CODE REVIEW

**Comment by parrt** [comment]
Wow! Interesting change. Let me consider this.

**Comment by parrt** [comment]
@sharwell do you think it's safe to add generics here? Is it worth it?  I.e., is there really a problem we're solving here?

**Comment by sharwell** [comment]
I think the concept is cool but I'd like to review it when I'm back at a computer to see if we can improve accuracy of some of the implementation details.

File: src/org/stringtemplate/v4/Interpreter.java [hunk 1]
```diff
@@ -793,7 +793,7 @@ protected int writePOJO(STWriter out, InstanceScope scope, Object o, String[] op
         String formatString = null;
         if ( options!=null ) formatString = options[Option.FORMAT.ordinal()];
         // ask the native group defining the surrounding template for the renderer
-        AttributeRenderer r = scope.st.impl.nativeGroup.getAttributeRenderer(o.getClass());
+        AttributeRenderer<? super Object> r = scope.st.impl.nativeGroup.getAttributeRenderer(o.getClass());
```
Comment (sharwell) [inline src/org/stringtemplate/v4/Interpreter.java]: 😟 I'm not a fan of this. It is definitely an inaccurate representation of the types involved. My recommendation is to extract the code that initializes `v` to a new helper method:

```
private <T> String renderObject(InstanceScope scope, String formatString, Object o, Class<T> attributeType) {
	AttributeRenderer<? super T> r = scope.st.impl.nativeGroup.getAttributeRenderer(attributeType);
	if ( r!=null ) {
		return r.toString(attributeType.cast(o), formatString, locale);
	} else {
		return o.toString();
	}
}
```

Then update this method to define and initialize `v` with the following:

```
String v = renderObject(scope, formatString, o, o.getClass());
```

File: src/org/stringtemplate/v4/StringRenderer.java [hunk 2]
```diff
@@ -40,11 +40,10 @@
  *  <li>{@code xml-encode}:</li>
  * </ul>
  */
-public class StringRenderer implements AttributeRenderer {
+public class StringRenderer implements AttributeRenderer<String> {
     // trim(s) and strlen(s) built-in funcs; these are format options
     @Override
-    public String toString(Object o, String formatString, Locale locale) {
-        String s = (String)o;
+    public String toString(String s, String formatString, Locale locale) {
```
[the remark this answers is not in the document]
Comment (Clashsoft) [reply src/org/stringtemplate/v4/StringRenderer.java]: Addressed by f0cacdb89d38e7d3037b6923c6ee00059a337fb2

File: src/org/stringtemplate/v4/STGroup.java [hunk 3]
```diff
@@ -147,6 +147,7 @@
      *  <p>
      *  This structure is synchronized.</p>
      */
+    @SuppressWarnings("rawtypes")
     protected Map<Class<?>, AttributeRenderer> renderers;
```
[the remark this answers is not in the document]
Comment (Clashsoft) [reply src/org/stringtemplate/v4/STGroup.java]: Unfortunately this causes an error in `getAttributeRenderer`:

Screenshot 2019-12-29 at 11 01 30

Adding a cast to `AttributeRenderer<? super T>` produces an unchecked warning. Seems like there is no way to implement it without any warnings :( I do agree that this is the overall cleaner solution.

File: src/org/stringtemplate/v4/Interpreter.java (hunk 1 above)
Comment (Clashsoft) [reply src/org/stringtemplate/v4/Interpreter.java]: Addressed by 7d870fae2f5c38180e3333228cc903a4d83a651c

File: src/org/stringtemplate/v4/STGroup.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Clashsoft) [reply src/org/stringtemplate/v4/STGroup.java]: Addressed by a53f0287b83dbf00a4f4eeff8c879f166606338d

File: src/org/stringtemplate/v4/STGroup.java [hunk 4]
```diff
@@ -754,17 +755,18 @@ public ModelAdaptor getModelAdaptor(Class<?> attributeType) {
...
 
-    public void registerRenderer(Class<?> attributeType, AttributeRenderer r, boolean recursive) {
+    public <T> void registerRenderer(Class<? extends T> attributeType, AttributeRenderer<? super T> r, boolean recursive) {
         if ( attributeType.isPrimitive() ) {
             throw new IllegalArgumentException("can't register renderer for primitive type "+
                                                attributeType.getSimpleName());
         }
 
         if ( renderers == null ) {
+            //noinspection rawtypes
```
[the remark this answers is not in the document]
Comment (Clashsoft) [reply src/org/stringtemplate/v4/STGroup.java]: Addressed by a53f0287b83dbf00a4f4eeff8c879f166606338d

**Comment by parrt** [comment]
Thanks alot, @sharwell !

**Comment by sharwell** [comment]
@Clashsoft would you be willing to submit a pull request to make the same change for `ModelAdaptor` after we merge this?

**Pull request merged**